server/irmaserver/sessions.go

```diff
@@ -138,58 +156,194 @@ func (s *memorySessionStore) deleteExpired() {
+	val, err := s.client.Get(context.Background(), sessionLookupPrefix+t).Result()
+	if err == redis.Nil {
+		return nil, nil
+	} else if err != nil {
+		_ = server.LogError(err)
+		return nil, errors.New("redis error")
+	}
+
+	var session session
+	session.conf = s.conf
+	session.sessions = s
+	if err := session.sessionData.UnmarshalJSON([]byte(val)); err != nil {
+		_ = server.LogError(err)
+		return nil, errors.New("redis error")
+	}
+	session.request = session.Rrequest.SessionRequest()
+
+	if session.LastActive.Add(maxSessionLifetime).Before(time.Now()) && !session.Status.Finished() {
+		s.conf.Logger.WithFields(logrus.Fields{"session": session.Token}).Infof("Session expired")
+		session.markAlive()
+		session.setStatus(server.StatusTimeout)
+		_ = s.update(&session) // Worst case the TTL and status aren't updated. We won't deal with this error
+	}
+
+	return &session, nil
+}
+
+func (s *redisSessionStore) add(session *session) error {
+	timeout := 2 * maxSessionLifetime // logic similar to memory store
+	if session.Status == server.StatusInitialized && session.Rrequest.Base().ClientTimeout != 0 {
+		timeout = time.Duration(session.Rrequest.Base().ClientTimeout) * time.Second
+	} else if session.Status.Finished() {
+		timeout = maxSessionLifetime
+	}
+
+	sessionJSON, err := session.sessionData.MarshalJSON()
+	if err != nil {
+		_ = server.LogError(err)
+		return errors.New("redis error")
+	}
+
+	err = s.client.Set(session.context, tokenLookupPrefix+session.sessionData.Token, session.sessionData.ClientToken, timeout).Err()
+	if err != nil {
+		_ = server.LogError(err)
+		return errors.New("redis error")
+	}
+	err = s.client.Set(session.context, sessionLookupPrefix+session.sessionData.ClientToken, sessionJSON, timeout).Err()
+	if err != nil {
+		_ = server.LogError(err)
+		return errors.New("redis error")
+	}
+
+	return nil
+}
+
+func (s *redisSessionStore) update(session *session) error {
+	return s.add(session)
+}
+
+func (s *redisSessionStore) stop() error {
+	err := s.client.Close()
+	if err != nil {
+		return server.LogError(err)
+	}
+
+	return nil
+}
+
+func (s *redisSessionStore) deleteExpired() {
+	// This method is not used as data is directly stored in Redis with a corresponding expiration time.
+}
+
 var one *big.Int = big.NewInt(1)
 
-func (s *Server) newSession(action irma.Action, request irma.RequestorRequest) *session {
+func (s *Server) newSession(action irma.Action, request irma.RequestorRequest, ctx context.Context) (*session, error) {
 	token := common.NewSessionToken()
 	clientToken := common.NewSessionToken()
 
 	base := request.SessionRequest().Base()
 	if s.conf.AugmentClientReturnURL && base.AugmentReturnURL && base.ClientReturnURL != "" {
 		if strings.Contains(base.ClientReturnURL, "?") {
-			base.ClientReturnURL += "&token=" + token
+			base.ClientReturnURL += "&Token=" + token
 		} else {
-			base.ClientReturnURL += "?token=" + token
+			base.ClientReturnURL += "?Token=" + token
```

Comment:
This is again an automatic refactoring mistake.

Side note: the fact that the unit tests don't detect this error imply to me that a test is missing for client return URLs with tokens (but this is Redis unrelated).